compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructure-array-declaration-to-context-var.js

```diff
@@ -6,7 +6,7 @@ function Component(props) {
     x = identity(props.value[0]);
   };
   foo();
-  return {x};
+  return <div>{x}</div>;
```

Comment:
why do we need to change here? this would force it to freeze, but why is that relevant?

Reply:
Ahh this is kind of an edge case. It's not necessary to freeze x before the return statement. I needed to add either jsx or a hook call to get `snap` to infer that this function is a `Component`|`Hook` (instead of `Other` type). This is not a problem internally as we rely on component syntax, so every function we attempt to compile is guaranteed to be a component or hook. It's also not an issue for `compilationMode: "infer"` cases without component syntax. Snap, however, uses `compilationMode: 'all'` by default. This mode should probably be patched to infer functions that don't create jsx / call hooks by function naming.

We filter by fnKind = 'Hook' | 'Component' when computing known immutable identifiers [here](https://github.com/facebook/react/blob/999bb6bebcd83570c936fe6f25f0b15de5000140/compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts#L298-L313)
Without this, we infer a (long) mutable range for `props` due to it being directly aliased into a context variable, which gets captured / mutated within a lambda.

Reply:
ahhh got it, thanks for clarifying!